url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/VO.java

```diff
@@ -0,0 +1,1445 @@
+import java.util.Queue;
+
+import mesh.IntArray;
+
+import Math.MathUtility;
+
+import spatialindex.core.IShape;
+import spatialindex.core.Point;
+import spatialindex.core.Region;
+import spatialindex.core.VOErrorException;
+import utility.StatisticForAuth;
+import utility.knnwithcryop;
+import utility.Compare.DataOfPoint;
+import utility.Compare.DistanceCompare;
+import utility.geo.DataOfLine;
+import utility.geo.Line;
+import utility.geo.VoronoiNeighbors;
+import utility.security.DataIO;
+import utility.security.Gfunction;
+import utility.security.Hasher;
+import utility.security.SecurityUtility;
+
+/**
+ * note that, if isSoloAuth = true, no privacy is considered. 
+ * @author chenqian
+ * */
+
+public class VO implements Serializable {
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = 1L;
+
+	private int queryHits;
+	private boolean Debug = false;
+	/**
+	 * use for counting in analyzing VO String
+	 */
+	private String rsaEntaValue;
+	private boolean isParallel = false;
+	private int ThreadNum = 16;
+	private HashMap<Integer, VOCell> nodeVO = new HashMap<Integer, VOCell>();
+	private HashMap<Integer, VOCell> dataVO = new HashMap<Integer, VOCell>();
+	private ArrayList<dcCell> dcVO = new ArrayList<dcCell>();
+	private ArrayList<lineCell> lineVO = new ArrayList<lineCell>();
+	private ArrayList<gfCell> gfVO = new ArrayList<gfCell>();
+	private RTree rtree;
+	private SecurityTree srtree;
+	private StatisticForAuth statForAuth;
+	private String sigWithRSA;
+	private int type_VO; // 0 means rtree based, 1 means voronoi diagram based, 2 means embeded kdtree rtree based
+	private int limit;
+	public static boolean isSoloAuth = true;
+	private ArrayList<StateCell> states;
+	public VO(){	
+	}
+	
+	void handleVOInParallel(final Point query, final int type){
+		final ArrayList<cell> vos = new ArrayList<VO.cell>();
+		for(int i = 0; i < dcVO.size(); i++){
+			vos.add(dcVO.get(i));
+		}
+		for(int i = 0; i < lineVO.size(); i++){
+			vos.add(lineVO.get(i));
+		}
+		for(int i = 0; i < gfVO.size(); i++){
+			vos.add(gfVO.get(i));
+		}
+		if(isParallel()){
+			final boolean[] threadStatus = new boolean[ThreadNum];
+			final int[] lock = new int[1];
+			//final int limit = vos.size();
+			lock[0] = 0;
+			for(int id = 0; id < ThreadNum; id ++){
+				threadStatus[id] = false;
+				final int tid = id;
+				new Thread(new Runnable() {
+					int threadId;
+					@Override
+					public void run() {
+						// TODO Auto-generated method stub
+						threadId = tid;
+						while(true){
+							int curid;
+							synchronized (lock) {
+								curid = lock[0];
+								lock[0] ++;
+							}
+							if(curid >= vos.size())break;
+							if(type == 0){
+								vos.get(curid).verify(query);
+							}
+							else vos.get(curid).generateVeryfyPart(query);
+						}
+						threadStatus[threadId] = true;
+					}
+				}).start();
+			}
+			while(true){
+				boolean found = false;
+				try {
+					Thread.sleep(100);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)